README.md
## Big "thank you" to MJ & Serena for this example.

[https://github.com/duncantl/MJSRandPerm](https://github.com/duncantl/MJSRandPerm)

# Current Status (Feb 22, 2022)

Currently, the modified version (see below) runs almost 7.5 times faster
than the original code for a `rpIter = 250`, the number of iterations in the loop.
This should also consume less memory (since the formula created in the pairTrials_RandomPerm() function
was capturing the call frame and the data.frame and was part of the return value.)


We could probably get more speed
  + by using lmerControl()
    + to pass good starting values 
  + avoiding some unnecessary checks.


# CAVEAT  

To get a little more speed, we modify three functions in the search path
+ stopifnot()
+ [[.data.frame
+ emmeans::.get.outside.method

The code modifies these just before running the final loop to simulate and fit the models.
At the end, the code resets them.
If there is an error during the loop, these functions will remain set to their new versions.
Call
```r
resetNSFunctions(origFuns)
```
or restart the R session.

## Files

### Original Files
+ origScript.R 
   + This was the starting point.
   + It contains 
       + 2 functions, 
	   + the 6 initialization steps
	   + the loop to create the permutation data frame and fit the model

+ origFuns.R
   + the 2 original functions extracted from origScript.R

+ origScript_noFuns.R
  + the origScript.R without the 2 functions moved to origFuns.R
  


### New Versions

We need each of the following 4 files to run the code

+ MJ&Serena_RandPermutationScript_20220202.R
   + the full script, modified from origScript_noFuns.R

+ funs.R
   + 2 functions from origFuns.R
   + significantly modified version of `pairTrials_RandomPerm()`
      + about a factor of 6-7 times faster overall
	     + 17 times faster for making the permutation data.frame.
   + induceMissingTrials() not changed since only called once.
      + could be made faster if important to do so.

+ loop.R
   + the loop (step 7) separated from the origScript.R
     + modified to move some computations outside of the loop

+ setNSFunctions.R
   + Two functions that set and reset functions in packages in the search path to get a little more
     speed.

+ vec.R 
   + two versions of a function to pair records - pairRecords1() and pairRecords2()
     + pairRecords2() is vectorized in the entire data.frame


### Sample Data Set

+ Sample0751-NCMeanAmpOutput.csv


### Questions and Notes

+ Questions.md 
  + High-level questions about objective and task
+ MetaExplore.md 
  + initial exploration of the code, some programmatically with meta-programming.

### Mechanism and data to identify C routines identified in profiler as .External2

These mostly come from the tidyverse code used in the original script 
and found when profiling the code.
They appear in the profling output as .External2 and it does not identify which
routine. So we traced these via the native debugger lldb and a Python script that
is called each time we hit the 

+ ExternalCalls.json
  + the actual calls and number of times they were called.
+ external.py
  + the python code that is called each time the do_External  C routine is called.
+ ext.c
  + a sample routine called via .External2

---


## Current Summary

+ We seem to have improved the creation of the data.frame for fitting each model by a factor of
   between 16 and 17. Probably more now when we use pairRecords2(). 

+ When we combine this with fitting the model, the speedup is a factor of about 7-8.
   + The improvement from creating the data.frame gets diminished as the rest of the code takes a
     non-trivial amount of time.


+ Profiling the entire script
```
                       self.time self.pct total.time total.pct
"[.data.frame"             32.92     4.76      55.90      8.09
"data.frame"               30.44     4.41      69.08     10.00
"validObject"              29.90     4.33      50.66      7.33
"<Anonymous>"              27.88     4.04     281.56     40.75
"genD.default"             25.44     3.68      54.74      7.92
"apply"                    25.10     3.63      58.70      8.50
"order"                    22.68     3.28      31.14      4.51
".getClassesFromCache"     21.84     3.16      21.88      3.17
".all.vars"                21.48     3.11      25.04      3.62
"model.frame.default"      20.98     3.04      49.16      7.11
".Call"                    19.90     2.88      22.80      3.30
"subbars"                  15.28     2.21      17.32      2.51
"$"                        14.24     2.06      52.94      7.66
".nextMethod"              14.04     2.03      57.62      8.34
"installClassMethod"       14.00     2.03      23.92      3.46
```



# Previous Profile Output

```
                       self.time self.pct total.time total.pct
"ls"                       11.92    18.67      11.94     18.70
"[.data.frame"              7.72    12.09      12.76     19.98
"stopifnot"                 4.50     7.05       4.54      7.11
"order"                     3.60     5.64       4.06      6.36
"rbind"                     3.54     5.54       4.04      6.33
"<Anonymous>"               1.52     2.38      34.30     53.71
"utils::getAnywhere"        1.52     2.38      15.08     23.61
".Call"                     1.34     2.10       1.52      2.38
".getClassesFromCache"      1.32     2.07       1.32      2.07
"data.frame"                1.24     1.94       3.62      5.67
"$"                         1.04     1.63       2.86      4.48
"initialize"                1.02     1.60       7.48     11.71
"%in%"                      0.94     1.47      13.50     21.14
"validObject"               0.86     1.35       1.90      2.98
".nextMethod"               0.82     1.28       2.46      3.85
"apply"                     0.80     1.25       2.24      3.51
"model.frame.default"       0.76     1.19       1.58      2.47
"jacobian.default"          0.60     0.94       1.26      1.97
"installClassMethod"        0.60     0.94       0.88      1.38
"getClassDef"               0.56     0.88       2.04      3.19
```

+ 2 iterations of the loop

+ collecting the call stacks for ls() just running loop.R, we have 289 calls to ls(), 9 unique call stacks .
   + 3 are happening in the byte-code compiler.
      + we can try to byte compile these ahead of time (e.g. put code in a package)
	    or just let this one-time events happen and diminish in signifance across more
		iterations of the loop.
   + 6 recover_data calling .get.outside.method("recover_data", cl),  "emm_basis"

   + debugging emm_basis, looking for emm_basis.lmerModLmerTest
      + doesn't find it so just calls UseMethod()
	  
   + Checking each of the 11 calls to  .get.outside.method,  we find all return NULL,
     so a complete waste of time in this case.
```r
val = vector("list", 300)
ctr = 1L
trace(emmeans:::.get.outside.method, exit = quote({val[[ctr ]] <<- returnValue(); ctr<<- ctr+1L}),
print = FALSE)
```
+ We replace .get.outside.method with a function that just returns NULL
```r
ns = getNamespace("emmeans")
unlockBinding(".get.outside.method", ns)
assign(".get.outside.method", function(generic, cls) NULL, ns)
```

+ Looking at stopifnot(). rpIter = 20 iterations of the loop
   + 2541 calls to stopifnot; 350 unique call stacks
   + srcfilecopy() called from source()
   + lmerControl() 
      + to check if object is a list.
	  + .makeCC x 3
   + We'll make stopifnot a no-op.  If we just return invisible(), we get an error
     in `KhatriRao(sm, t(mm)) : 'list' object cannot be coerced to type 'integer'`
	 So we force the evaluation of the arguments as this is probably lazy evaluation.
```r
old.stopifnot = stopifnot
ns = getNamespace("base")
unlockBinding("stopifnot", ns)
assign("stopifnot", function (..., exprs, exprObject, local = TRUE) { list(...); if(!missing(exprs)) exprs; if(!missing(exprObject)) exprObject;  invisible()}, ns)
```

```r
f = `[[.data.frame`
body(f) = body(f)[-3]
ns = getNamespace("base")
unlockBinding("[[.data.frame", ns)
assign("[[.data.frame", f, ns) 
```

   + See setNSFunctions.R for functions to set and reset these.
     + Currently, if there is an error between setting and resetting these, they will be left in their modified state.
	   Call `resetNSFunctions(origFuns)` or restart the R session.

+ With .get.outside.method and stopifnot made degenerate, we get
```
                      self.time self.pct total.time total.pct
"[.data.frame"             0.50    12.95       0.90     23.32
"order"                    0.34     8.81       0.38      9.84
"rbind"                    0.32     8.29       0.36      9.33
"<Anonymous>"              0.20     5.18       1.58     40.93
".Call"                    0.16     4.15       0.16      4.15
"data.frame"               0.10     2.59       0.32      8.29
"validObject"              0.10     2.59       0.16      4.15
".identC"                  0.10     2.59       0.10      2.59
"initialize"               0.08     2.07       0.60     15.54
"apply"                    0.08     2.07       0.22      5.70
"model.frame.default"      0.08     2.07       0.20      5.18
".all.vars"                0.08     2.07       0.12      3.11
"genD.default"             0.06     1.55       0.24      6.22
"$"                        0.06     1.55       0.18      4.66
".adj.p.value"             0.06     1.55       0.06      1.55
".class1"                  0.04     1.04       0.14      3.63
"paste"                    0.04     1.04       0.10      2.59
"%in%"                     0.04     1.04       0.06      1.55
"getClassDef"              0.04     1.04       0.06      1.55
"jacobian.default"         0.04     1.04       0.06      1.55
```

+ Running
```r
rpIter = 250
e = new.env()
source("origFuns.R", e)
z = e$pairTrials_RandomPerm(dfMissing) # evaluate multiple times to byte compile
tm.orig = system.time(source("origScript_noFuns.R", e))

tm.new = system.time(source("MJ&Serena_RandPermutationScript_20220202.R"))

tm.orig/tm.new
    user   system  elapsed 
6.627625 4.354592 6.603736 

rpIter = 1000

    user   system  elapsed 
6.432836 4.260328 6.413952 
```

```
tm.orig
   user  system elapsed 
308.521   1.589 311.630 
```
311.630            250
873.38 seconds for 500 iterations.
1374.47           1000

d = data.frame(tmOrig = c(311.63, 656.38, 1374.47), N = c(250, 500, 1000), tmNew = c(41.557 , 75.795, 172.752),
                tmPair2 = c(36.189, 77.042, 182.349))

+ With rbindlist rather than rbind, for rpIter = 250
```
tm.orig/tm.new
    user   system  elapsed 
7.539982 4.226064 7.498857 
```

   
## Some Suggestions/Questions

See [Changes.md](Changes.md) for some of the very high-level changes related
to 
  + changing computations done in loops and moving them outside and after the loop in a vectorized
    manner
  + suggestions for simplifying/removing  code that is redundant

Most of these current changes are in the updated script
[MJ&Serena_RandPermutationScript_20220202.R](MJ&Serena_RandPermutationScript_20220202.R).


## Timing the Original and New Script

The original script (with the number of iterations adjusted)
```r
source("origScript.R")
```

The modified version can be run 
```r
source("funs.R")
source("MJ&Serena_RandPermutationScript_20220202.R")
```

We probably have to run each of these several times to get the byte-code compilation.
This is a reason for separating the functions from origScript.R so that we
are not redefining them each time we source that script.


Profiling data from the modified version that doesn't use the tibble or tidyverse functions.
There 
```
                      self.time self.pct total.time total.pct
"[<-.data.frame"           0.06    16.67       0.06     16.67
"exists"                   0.06    16.67       0.06     16.67
"withCallingHandlers"      0.06    16.67       0.06     16.67
".Call"                    0.02     5.56       0.18     50.00
"DataMask$new"             0.02     5.56       0.08     22.22
"findCenvVar"              0.02     5.56       0.08     22.22
"group_data"               0.02     5.56       0.06     16.67
"enexpr"                   0.02     5.56       0.02      5.56

Changes.md
+ Most likely in the lme4/nlme fitting, but maybe generating the permutations.
+ Looks like can parallelize this. So get on a large machine and run in parallel.

+ Within the loop (which one?) , there is a loop from 1:ageN and then set the first two names names of the result to
  age and estimate. Is the same value of 2 for  ageN = 2 and setting the 2 names a coincidence or
  supposed to be connected.  Probably a coincidence.

+ Rather than append  LMEMis_output_RP_oneIter to LMEMis_output_RP_allIter in each iteration,
   + pre-allocate and fill in rows
   + keep as a list until the very end and then use do.call(rbind, LMEMis_output_RP_allIter)


+ Set the names later, i.e., outside the loop after it has complete.
  + Same with computing derived variables on columns in the final data.frame.

+ Do the between at the end in vectorized form. 2 steps for the young and old corresponding to the two
  values of ageArray.
  + Do outside of the loop, after it has completed


+ dfMissing_pairedWide_RP is assigned but never used in each iteration.
  + And this involves the list[] which probably has a non-trivial overhead and uses non-standard evaluation.
  + So we can simplify this in the loop with
```
  fit.LMEMis_RP <- pairTrials_RandomPerm(dfMissing)[[2]]
```
  + Can also think about simplifying the function and not computing/returning the first element
  
+ Good to put names on the elements of the list returned by pairTrials_RandomPerm
  + this allows us to use 
  `pairTrials_RandomPerm(dfMissing)$fit`
   rather than `pairTrials_RandomPerm(dfMissing)[[2]]`


+ These changes seem to ADD 10 seconds, i.e., for rpIter = 100
   + origScript.R takes 130 seconds
   + the modified script (commit 10264bb4a) takes 140.
   + Not sure these numbers are reliable, and they should change when we do a lot more iterations.


+ Calling complete.cases() at the start of each call to pairTrials_RandomPerm()
   but the input is the same for each iteration of the loop calling this. 
   + So can remove the NA rows once and pass this as the second argumet to pairTrials_RandomPerm()


+ for(i in 1:rpIter)  is ostensibly faster than replicate(), in this situation !
  + 11.1-11.5 seconds for 10 iterations, versus 12.9-13.2
  + may be different for many iterations.


+ ¿ Remove emmeans:: qualifier for speed ?











##
induceMissingTrials() seems to be only called once.



## Profiling

.External2 takes the most time, but a very flat/dispersed profiling table with nothing taking 
significantly most of the time.

To see what .External2 calls are taking the longest we can use lldb, or simpler, 
look at the corresponding lines in the Rprof output.
```
ll = readLines("tm3.prof")
ex2 = grep('^".External2"', ll)
ex.funs = strsplit(ll[ex2], " ")
table(sapply(ex.funs, `[`, 2))
```

Similarly, we can identify the source of the FUN in the apply functions with
```r
table(sapply(strsplit(ll[grep('^"FUN"', ll)], " "), `[`, 3))
```


### ls()

```r
ls.stack = collectCallStacks(ls)
box = new.env(); system.time(source("MJ&Serena_RandPermutationScript_20220202.R", box))
ls.calls = ls.stack()
length(ls.calls)
```
1504 calls for 10 iterations.

table(sapply(ls.calls, length))

 14  19  21  30  62 
600 600 300   2   2 





Using lldb to print out the symbol in each call to .External

ll = readLines("lldb.external.output")
routines = ll[grep("^Command #2", ll) + 1]
dsort(table(routines))

This is for 10 iterations in the original script.
```
routines
    vctrs_new_data_frame                  vctrs_c 
                   22310                    19600 
        rlang_ext2_call2 rlang_ext_capturearginfo 
                   10000                     8000 
         rlang_ext2_exec          rlang_ext2_eval 
                    8000                     5100 
   rlang_ext_dots_values              vctrs_cbind 
                    4000                     1500 
    vctrs_recycle_common     rlang_ext_arg_match0 
                    1100                      700 
             C_compcases              vctrs_rbind 
                     521                      510 
           magrittr_pipe     rlang_ext2_eval_tidy 
                     500                      500 
             C_termsform        vctrs_cast_common 
                     110                      100 
       vctrs_type_common             C_modelframe 
                     100                       81 
           C_modelmatrix 
                      50 
```


Based on a single iteration of the original script, we have
```
             "C_termsform"             "C_modelframe" 
                        12                          9 
             "C_compcases"     "vctrs_new_data_frame" 
                        53                       2951 
"rlang_ext_capturearginfo"          "rlang_ext2_eval" 
                      1700                        600 
    "rlang_ext_arg_match0"        "vctrs_type_common" 
                       250                        100 
       "vctrs_cast_common"                  "vctrs_c" 
                       100                       2050 
    "vctrs_recycle_common"            "magrittr_pipe" 
                       200                         50 
        "rlang_ext2_call2"    "rlang_ext_dots_values" 
                      1000                        400 
         "rlang_ext2_exec"     "rlang_ext2_eval_tidy" 
                       800                         50 
             "vctrs_cbind"              "vctrs_rbind" 
                       150                         51 
           "C_modelmatrix" 
                         5 
```



After profiling and seeing .External2 as the most time consuming function,
we used LLDB to get the names of the routines being called via .External2.
That shows a lot of vctrs, rlang, magrittr_pipe (see Changes.md).
So where are these coming from. Let's look at the functions used in
pairTrials_RandomPerm.


```
g = getGlobals(pairTrials_RandomPerm)
gfuns = unique(g$functions)
where = sapply(gfuns, find)

origScript.R
# MJ and Serena script: Randomly pairing trial data and then fitting a
# linear mixed effects model

# Random permutation function starts on lines 160-214
# The random permutation function is called 10,000 times on line308

# This script imports a simulated mean amplitude file and induces missing trials
# based on the specified missingness pattern (e.g., more missing data in later 
# trials and in younger subjects) and percentage of subjects with low trial-count 
# (e.g., 6% of subjects are induced to have less than 10 trials/condition 
# remaining). Then, trial pairings are calculated by randomly permuting
# the dataset (10,000 times per dataset). Finally, a linear mixed effects (LME) 
# model is fitted to each randomly permuted trial-level dataset. 

# Requirements: 
  # - filename: One simulated sample's .csv file containing the following columns, 
  #   which are labelled based on the convention that lowercase variables describe 
  #   fixed effects (e.g., emotion) and capital-letter variables describe random 
  #   effects (e.g., SUBJECTID):
    # - SUBJECTID: Simulated subject ID (e.g., 01, 02, ...)
    # - age: Simulated age group (i.e., youngerAgeGroup, olderAgeGroup)
    # - emotion: Simulated emotion condition (i.e., A, B)
    # - ACTOR: Simulated stimulus actor ID (i.e., 01, 02, 03, 04, 05)
    # - presentNumber: Presentation number of specific stimulus (emotion 
    #   condition/actor) ranging from 1 to 10
    # - meanAmpNC: Simulated mean amplitude value (in units of microvolts)
  # - See instructions in steps 5 and 6 for specifying the missingness pattern 
  #   (more missing data in later trials and/or in younger subjects or data
  #   missing completely at random (MCAR)) and percentage of subjects with low
  #   trial-count, respectively

# Script Functions:
  # 1. Define function for inducing missing trials
  # 2. Load simulated data file
  # 5. Specify missingness pattern 
  # 6. Induce missing data based on specified missingness pattern and percentage
  #    of subjects with low trial-count
  # 7. Fit LME model with trial-level dataset after inducing trial missingness

# Outputs: 
  # - Estimated marginal means for each age group and model (LME)
  #   for a dataset that has been randomly permuted 10,000 times 

library(plyr) # V.1.8.6; ddply function
library(lme4) # V.1.1-25; used for creating lme models
library(lmerTest) # V.3.1-3; used for returning p-value for lme models
library(gsubfn) # V.0.7; list function for assigning multiple outputs
library(data.table) # V.1.13.2; fread function
library(dplyr) # V.1.0.2; select function
library(performance) # V.0.6.1; check_convergence function
library(afex) # V.0.28-1; ANOVA analysis
library(emmeans) # V.1.5.3; extract estimated marginal means
library(car) # V.3.0-10; contr.sum function
library(stringr) # V.1.4.0; str_sub function
library(MatchIt) # V.4.1.0; used for pairing functions
library(tidyr) # V.1.1.2; used for spread function
library(ggplot2) # V.3.3.2; used for plotting

#-----------------------------------------------------------------------
# 1. DEFINE FUNCTION FOR INDUCING MISSING TRIALS

# induceMissingTrials: Function to randomly select subjects for inducing
# low trial counts and subsequent casewise deletion prior to ANOVA analysis. 
# In addition, missing trials are induced based on the specified probability
# weights from step 5 below. 
# - Format: 
#     list[dfMissing, subjectCaseDeletion, trialCount] <- induceMissingTrials(dfOriginal, caseDeletionPct) 
# - Inputs:
  # - dfOriginal: Dataframe with the simulated "population" dataset before any
  #   induced missing trials (see Outputs section at the top of the script for
  #   more information). 
  # - caseDeletionPct: Percent of subjects with low trial-count (i.e., less than
  #   10 trials/condition).
# - Outputs: List containing three elements:
  # - dfMissing: A copy of the dfOriginal after missing trials have been induced. 
  #   Rows with missing trials have a meanAmpNC value of NA.
  # - subjectCaseDeletion: Array of subject IDs that have been randomly selected
  #   for low trial counts.
  # - trialCount: Long dataframe listing the remaining number of trials per
  #   subject and emotion condition after inducing missing trials. It contains the 
  #   following columns: SUBJECTID, emotion, trialN (see Outputs section at the
  #   top of the script for more information). 
induceMissingTrials <- function(dfOriginal, caseDeletionPct) {
  # Create copy of the dfOriginal dataframe for inducing missing trials
  dfMissing <- data.frame(dfOriginal)
  
  # Extract age probability weights (one weight per subject)
  dfAgeWeight <- aggregate(ageWeight ~ SUBJECTID, dfOriginal, mean, 
                           na.action = na.omit)
  
  # Calculate the number of low trial-count subjects by multiplying caseDeletionPct
  # by the total number of subjects. If this value is not an integer, the output
  # is rounded up. 
  caseDeletionN <- ceiling((caseDeletionPct/100)*subjectN)
  
  # Randomly sample the subject IDs that will have low trial counts based on the 
  # specified age weights and the caseDeletionN variable
  subjectCaseDeletion <- sample(dfAgeWeight$SUBJECTID, caseDeletionN, 
                                replace = FALSE, prob=dfAgeWeight$ageWeight)
  
  # Calculate the maximum number of trials that can be removed from each condition
  # before the subject is considered to have a low trial count and would be
  # casewise deleted (e.g., if there are 50 trials, a maximum of 40 trials can
  # be removed for an included (not casewise deleted) subject)
  trialMissingThreshold <- emotionTrialN - 10
  
  # Loop through each subject and randomly select a subset of trials from each
  # condition to remove
  for (subject in dfAgeWeight$SUBJECTID) {
    
    # Generate the number of missing trials to induce for each emotion condition
    if (subject %in% subjectCaseDeletion) { 
      
      # For subjects with a low trial count, at least one emotion condition will 
      # have less than 10 trials remaining
      trialMissing <- c(sample(x=(trialMissingThreshold+1):emotionTrialN, size = 1),
                        sample(x=0:emotionTrialN, size = emotionN-1, replace = TRUE))                       
    } else { 
      # For subjects who do NOT have a low trial count, all emotion conditions
      # will have at least 10 trials
      trialMissing <- sample(x=0:trialMissingThreshold, size = emotionN, 
                             replace = TRUE)
    }
    
    # Shuffle order of emotion conditions and then loop through each condition
    # (This line is added so that one condition does not consistently have more
    # missing trials.)
    emotionLabelRand <- sample(emotionLabel)
    for (j in 1:length(emotionLabelRand)) {
      
      emotionTrialMissing <- trialMissing[j] # Extract the number of missing trials for this condition
      
      if (emotionTrialMissing != 0) { # If this emotion condition was not selected to have 0 missing trials
        
        # Find this subject and condition's rows in the dataframe
        subjectIndex <- which(dfMissing$SUBJECTID==subject & dfMissing$emotion==emotionLabelRand[j])
        
        # Extract the presentation number probability weights for this subject and condition
        subjectProbWeight <- dfMissing$presentNumberWeight[subjectIndex]
        
        # Randomly select the missing trials based on the specified probability 
        # weights and the number of missing trials
        subjectIndexMissing <- sample(subjectIndex, emotionTrialMissing, 
                                      replace = FALSE, prob=subjectProbWeight)
        
        # For these missing trials only, replace the meanAmpNC value with NA
        dfMissing[subjectIndexMissing,]$meanAmpNC <- NA
      }
    }
  }
  
  # Save the number of trials remaining for each subject and condition in a dataframe
  trialCount <- ddply(dfMissing, .(SUBJECTID, emotion), summarize, 
                      trialN = sum(!is.na(meanAmpNC)))
  names(trialCount)[3] <- 'trialN' # Update column name to trialN
  
  return(list(dfMissing, subjectCaseDeletion, trialCount)) # Return output variables
}

# pairTrials_RandomPerm: Function for pairing trials with exact match only
# - Format: 
#     list[dfMissing_pairedWide, fit.LMEMis] <- pairTrials_RandomPerm(dfMissing)
# - Inputs:
# - dfMissing: A copy of the dfOriginal after missing trials have been induced. 
#   Rows with missing trials have a meanAmpNC value of NA.
# - Output:
# - dfMissing_pairedWide: A wide dataframe with trials paired randomly. Missing trial pairings are designated with NA
pairTrials_RandomPerm <- function(dfMissing) {
  dfMissing_NoNA <- dfMissing[complete.cases(dfMissing), ] # Remove all NA rows (otherwise function will say that you've matched all pairs)
  
  # Initialize variable for storing paired data
  dfMissing_pairedWide <- NULL
  
  for (subject in unique(dfMissing_NoNA$SUBJECTID)) { 
    # Subset dfMissing_NoNA dataframe by subject
    dfMissing_NoNA_SubjectSubset <- dfMissing_NoNA[dfMissing_NoNA$SUBJECTID == subject,]
    
    # Find each emotion condition's rows
    emotionRows_A <- dfMissing_NoNA_SubjectSubset$emotion == "A"
    emotionRows_B <- dfMissing_NoNA_SubjectSubset$emotion == "B"

        # We sample both conditions A and B randomly so we get a random selection of actor/presentation number for both conditions
    # Otherwise, selecting 1:n may result in always pairing Actor 1/Trial 1 of Emotion A with a random trial from Emotion B
    dfMissing_NoNA_SubjectSubset$subclass[emotionRows_A] <- sample(1:sum(emotionRows_A))
    dfMissing_NoNA_SubjectSubset$subclass[emotionRows_B] <- sample(1:sum(emotionRows_B))
    
    # Convert dataframe with paired trials from long to wide
    dfMissing_pairedWide_SubjectSubset <- dfMissing_NoNA_SubjectSubset %>% 
      pivot_wider(names_from = emotion, names_sep = ".", values_from = c(meanAmpNC, ACTOR, presentNumber, presentNumberWeight))
    
    # If this subset does not have trials from both emotion conditions
    if (!("meanAmpNC.B" %in% colnames(dfMissing_pairedWide_SubjectSubset) & "meanAmpNC.A" %in% colnames(dfMissing_pairedWide_SubjectSubset))) { 
      next
    }
    
    # Calculate difference in amplitude between trial pairs
    dfMissing_pairedWide_SubjectSubset$meanAmpNC_BMinusA <- dfMissing_pairedWide_SubjectSubset$meanAmpNC.B - dfMissing_pairedWide_SubjectSubset$meanAmpNC.A
    
    # Remove all rows with NA
    dfMissing_pairedWide_SubjectSubset_NoNA <- dfMissing_pairedWide_SubjectSubset[complete.cases(dfMissing_pairedWide_SubjectSubset), ] # Remove all NA rows (otherwise function will say that you've matched all pairs)
    
    # Store this subject's paired trial dataframe in dfMissing_pairedWide
    dfMissing_pairedWide <- bind_rows(dfMissing_pairedWide, dfMissing_pairedWide_SubjectSubset_NoNA)
  }
  
  # Calculate average presentation number for each trial pair
  dfMissing_pairedWide$presentNumberAvg <- (dfMissing_pairedWide$presentNumber.A+dfMissing_pairedWide$presentNumber.B)/2

  # Fit LME model
  fit.LMEMis <- lmer(meanAmpNC_BMinusA ~ age + presentNumberAvg + (1|SUBJECTID), data=dfMissing_pairedWide, REML = TRUE)
  
  return(list(dfMissing_pairedWide, fit.LMEMis)) # Return dataframe with paired trials
  
}

#-----------------------------------------------------------------------
# 2. LOAD SIMULATED DATA FILE

# Specify filepath of simulated data file
#filename <- 'T:/Projects/InProgress/LME_DifferenceWaves&LatencyERP/Data/01_SEREEGASimulation/MatlabAndRCode/00_SampleCodeForTestingDifferenceWavePairing/Sample0751-NCMeanAmpOutput.csv'
filename = "Sample0751-NCMeanAmpOutput.csv"


# Import data sample ("population" dataset)
dfOriginal <- fread(filename) 

# Specify desired columns as factors for subsequent analysis 
dfOriginal$ACTOR <- as.factor(dfOriginal$ACTOR)
dfOriginal$SUBJECTID <- as.factor(dfOriginal$SUBJECTID)
dfOriginal$emotion <- as.factor(dfOriginal$emotion)
dfOriginal$age <- as.factor(dfOriginal$age)

#-----------------------------------------------------------------------
# 5. SPECIFY MISSINGNESS PATTERN 

set.seed(20210329) # Specify seed for reproducible results

# Specify probability weight distribution for presentation numbers 6-10 vs.
# 1-5. These weights (i.e., presentNumberWeight6to10 and presentNumberWeight1to5)
# sum to 1 and are used to specify missingness pattern for the within-subjects 
# effect.  
# - For example, if presentNumberWeight6to10 = 0.7 and presentNumberWeight1to5 = 0.3,
#   then 70% of missing trials are from presentation numbers 6-10 and 30% of 
#   trials are from presentation numbers 1-5.
# - If both weight variables are equal to 0.5, an equal number of missing trials 
#   are drawn from each presentation number (MCAR missingness).
presentNumberWeight6to10 <- 0.7
presentNumberWeight1to5 <- 1-presentNumberWeight6to10

# Calculate the total number of trials per condition for each group of presentation
# numbers (i.e., 6-10 and 1-5). This value is used to scale each individual trial's 
# presentation number weight so that the weights will sum to 1 (see lines 292-294). 
emotionTrialN <- length(unique(dfOriginal$ACTOR)) * length(unique(dfOriginal$presentNumber))  
presentNumberTrials6to10 <- emotionTrialN/2 
presentNumberTrials1to5 <- emotionTrialN/2 

# Specify probability weight distribution for younger vs. older age group. These 
# weights are used to specify missingness pattern for the between-subjects
# effect (e.g., if ageWeightYounger = 0.7, then 70% of subjects selected for more
# missing trials and subsequent casewise deletion were from the younger age group).
ageWeightYounger <- 0.7
ageWeightOlder <- 1-ageWeightYounger

# Calculate the total number of subjects in the younger and older age groups. This
# value is used to scale each subject's age weight so that the weights will sum to 
# 1 (see lines 295-296).
subjectN <- length(unique(dfOriginal$SUBJECTID))
ageTrialsYounger <- subjectN/2
ageTrialsOlder <- subjectN/2

# Create probability weight columns for presentation number and age group based
# on values specified above
dfOriginal$presentNumberWeight <- ifelse(dfOriginal$presentNumber > 5,
                                         (presentNumberWeight6to10/presentNumberTrials6to10), 
                                         (presentNumberWeight1to5/presentNumberTrials1to5))
dfOriginal$ageWeight <- ifelse(dfOriginal$age == 'youngerAgeGroup', 
                               ageWeightYounger/ageTrialsYounger, ageWeightOlder/ageTrialsOlder)


#-----------------------------------------------------------------------
# 6. INDUCE MISSING DATA BASED ON SPECIFIED MISSINGNESS PATTERN AND 
#    PERCENTAGE OF SUBJECTS WITH LOW TRIAL-COUNT

# Define variables needed for induceMissingTrials function
emotionLabel <- c("A", "B") # Name of each emotion condition
emotionN <- length(emotionLabel) # Number of emotion conditions

# Specify percent of subjects with low trial-count who will be casewise deleted
caseDeletionPct <-  11

# Use induceMissingTrials function to remove trials based on specified missingness
# (step 5) and caseDeletionPct variable
list[dfMissing, subjectCaseDeletion, trialCount] <- induceMissingTrials(dfOriginal,
                                                                        caseDeletionPct) 

#------------------------------------------------------------------------
# 7. PAIR TRIALS AND FIT LME MODEL WITH TRIAL-LEVEL DATASET AFTER INDUCING TRIAL MISSINGNESS

# In our full script, we call pairTrials_RandomPerm 120,000,000 times 
# (=10,000 permutations per dataset * 1,000 datasets * 4 types of induced missing trials for each dataset * 3 patterns of missing data)

# In this example, we have one dataset. We randomly permute this dataset 10,000 times,
# store emmeans results for each iteration in separate array, then calculate final emmeans summary
if(!exists("rpIter"))
    rpIter = 250 # was originally uncondition and 10000

presentAvgValue = 5.5 # Value needed for emmeans calculations
ageN <- 2 # Value needed for extracting rows from emmeans output
ageArray <- c(-1.998, 0) # Array needed for extrating emmeans output 

LMEMis_output_RP_allIter <- NULL # Temporary variable for storing emmeans for each random perm iteration

for (i in 1:rpIter) {
  # Pair trials with random permutation function
  list[dfMissing_pairedWide_RP, fit.LMEMis_RP] <- pairTrials_RandomPerm(dfMissing)
  
  # Extract marginal means from LME model
  mLME <- emmeans::emmeans(fit.LMEMis_RP, pairwise~age, mode = "satterthwaite", 
                           lmerTest.limit = 240000, at = list(presentNumberAvg = c(presentAvgValue)))
  
  # Extract output values for each emotion condition 
  LMEMis_output_RP_oneIter <- data.frame(summary(mLME, infer = c(TRUE, TRUE))$emmeans)
  # Check if each true population emotion condition value is located within the
  # model's 95% confidence interval for the emotion condition
  for (ageNum in 1:ageN) {
    LMEMis_output_RP_oneIter$inCL[ageNum] <- between(ageArray[ageNum], 
                                                     LMEMis_output_RP_oneIter$lower.CL[ageNum], 
                                                     LMEMis_output_RP_oneIter$upper.CL[ageNum])
  }
  names(LMEMis_output_RP_oneIter)[1] <- c('age') # Update column names
  names(LMEMis_output_RP_oneIter)[2] <- c('estimate')
  
  LMEMis_output_RP_oneIter$iteration <- i
  LMEMis_output_RP_allIter <- bind_rows(LMEMis_output_RP_allIter, LMEMis_output_RP_oneIter)
}

# We then extract summary values from LMEMis_output_RP_allIter
